File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        HashMap<Character,Integer> hm=new HashMap<>();
        for(char c:s1.toCharArray())
            hm.put(c,hm.getOrDefault(c,0)+1);
        int i=0,j=0;
        int count=hm.size();
        while(j<s2.length()){
            char c=s2.charAt(j);
            if(hm.containsKey(c)){
              hm.put(c,hm.getOrDefault(c,0)-1);
              if(hm.get(c)==0)
                count--;
              
            }
             while(count==0){
                  if(j-i+1==s1.length())
                      return true;
                    if(hm.containsKey(s2.charAt(i))){
                           hm.put(s2.charAt(i),hm.getOrDefault(s2.charAt(i),0)+1);
                           if(hm.get(s2.charAt(i))>0)
                               count++;
                    }
                    i++;
           }
              
             
            // while(map.getOrDefault(c,0)>hm.getOrDefault(c,0)){
            //     map.put(s2.charAt(i),map.get(s2.charAt(i))-1);
            //     if(map.get(s2.charAt(i))==0)
            //         map.remove(s2.charAt(i));
            //      i++;
            // }
            // if(hm.equals(map))
            //     return true;
            j++;
        }
        return false;
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.cs
public class Solution {
    public bool CheckInclusion(string s1, string s2) {
        Dictionary<char,int> dict1=new Dictionary<char,int>();
        foreach(char c in s1.ToCharArray()){
            if (dict1.ContainsKey(c)) {
                dict1[c]++;
            } else {
                dict1[c] = 1;
            }
        }
        int i=0,j=0;
        Dictionary<char,int> dict2=new Dictionary<char,int>();
        while(j<s2.Length){
            char c=s2[j];
            if (dict2.ContainsKey(c)) {
                dict2[c]++;
            } else {
                dict2[c] = 1;
            }
            while(dict2.GetValueOrDefault(c,0)>dict1.GetValueOrDefault(c,0)){
                char p=s2[i];
                dict2[p]--;
                if(dict2[p]==0)
                    dict2.Remove(p);
                i++;
                
            }
            if(AreDictionariesEqual(dict1,dict2))
                return true;
            j++;
        }
        return false;
    }
    bool AreDictionariesEqual(Dictionary<char, int> dict1, Dictionary<char, int> dict2) {
        if (dict1.Count != dict2.Count) return false;

        foreach (var kvp in dict1) {
            if (!dict2.ContainsKey(kvp.Key) || dict2[kvp.Key] != kvp.Value) {
                return false;
            }
        }

        return true;
    }
}